Source/Scene/ShadowMap.js

```diff
@@ -0,0 +1,397 @@
+        '../Renderer/PassState',
+        '../Renderer/PixelDatatype',
+        '../Renderer/RenderState',
+        '../Renderer/ShaderProgram',
+        '../Renderer/ShaderSource',
+        '../Renderer/Texture',
+        './Camera',
+        './CullFace',
+        './OrthographicFrustum',
+        './Pass',
+        './PerInstanceColorAppearance',
+        './Primitive'
+    ], function(
+        BoundingRectangle,
+        BoxOutlineGeometry,
+        Cartesian3,
+        Color,
+        ColorGeometryInstanceAttribute,
+        defined,
+        destroyObject,
+        GeometryInstance,
+        Matrix4,
+        PixelFormat,
+        SphereOutlineGeometry,
+        ClearCommand,
+        Framebuffer,
+        PassState,
+        PixelDatatype,
+        RenderState,
+        ShaderProgram,
+        ShaderSource,
+        Texture,
+        Camera,
+        CullFace,
+        OrthographicFrustum,
+        Pass,
+        PerInstanceColorAppearance,
+        Primitive) {
+    "use strict";
+
+    // TODO : doc
+    // TODO : handle if depth texture extension is not supported
+
+    /**
+     * @private
+     */
+    function ShadowMap(scene) {
+        this.enabled = true;
+
+        this.debugShow = false;
+        this.debugFreezeFrame = false;
+        this._debugLightFrustum = undefined;
+        this._debugLight = undefined;
+        this._debugCameraFrustum = undefined;
+        this._debugCamera = undefined;
+        this._debugShadowViewCommand = undefined;
+
+        this._depthStencilTexture = undefined;
+        this._framebuffer = undefined;
+        this._size = 1024; // Width and height of the shadow map in pixels
+
+        this.renderState = RenderState.fromCache({
+            cull : {
+                enabled : true,
+                face : CullFace.BACK
+            },
+            depthTest : {
+                enabled : true
+            },
+            colorMask : {
+                red : false,
+                green : false,
+                blue : false,
+                alpha : false
+            },
+            depthMask : true,
+            viewport : new BoundingRectangle(0, 0, this._size, this._size)
+        });
+
+        this.passState = new PassState(scene.context);
+
+        // TODO : determine frustum based on scene
+        var frustum = new OrthographicFrustum();
+        frustum.left = -15.0;
+        frustum.right = 15.0;
+        frustum.bottom = -15.0;
+        frustum.top = 15.0;
+        frustum.near = 1.0;
+        frustum.far = 100.0;
+        this._frustum = frustum;
+
+        // TODO : position camera based on sun direction
+        var camera = new Camera(scene);
+        camera.frustum = frustum;
+        var centerLongitude = -1.31968;
+        var centerLatitude = 0.698874;
+        camera.lookAt(Cartesian3.fromRadians(centerLongitude, centerLatitude), new Cartesian3(0.0, 0.0, 75.0));
+        this.camera = camera;
+
+        this.clearCommand = new ClearCommand({
```

Comment:
A better vertical code ordering would be to move this to just under where the render state is created.